packages/gestalt/src/Tooltip.js

```diff
@@ -0,0 +1,81 @@
+// @flow
+
+import * as React from 'react';
+import Controller from './Controller.js';
+import Text from './Text.js';
+import Box from './Box.js';
+
+const noop = () => {};
+
+function Popout({ anchor, text }: {| anchor: ?HTMLElement, text: string |}) {
+  if (!anchor) {
+    return null;
+  }
+
+  return (
+    <Controller
+      anchor={anchor}
+      bgColor="darkGray"
+      caret={false}
+      idealDirection="down"
+      onDismiss={noop}
+      positionRelativeToAnchor
+      size={null}
+    >
+      <Box maxWidth={180} paddingY={1} paddingX={2} role="tooltip">
+        <Text color="white" size="xs">
+          {text}
+        </Text>
+      </Box>
+    </Controller>
+  );
+}
+
+type Props = {|
+  children: React.Node,
+  text: string,
+|};
+
+type State = {|
+  focused: boolean,
+  hovered: boolean,
+|};
+
+export default class Tooltip extends React.Component<Props, State> {
+  state = {
+    focused: false,
+    hovered: false,
+  };
+
+  childRef = React.createRef();
+
+  handleBlur = () => this.setState({ focused: false });
+
+  handleFocus = () => this.setState({ focused: true });
+
+  handleMouseEnter = () => this.setState({ hovered: true });
+
+  handleMouseLeave = () => this.setState({ hovered: false });
+
+  render() {
+    const { children, text } = this.props;
+    const { focused, hovered } = this.state;
+
+    return (
+      <Box display="inlineBlock">
+        <Box
+          onBlur={this.handleBlur}
+          onFocus={this.handleFocus}
+          onMouseEnter={this.handleMouseEnter}
+          onMouseLeave={this.handleMouseLeave}
+          ref={this.childRef}
+        >
+          {children}
+        </Box>
+        {(hovered || focused) && (
+          <Popout anchor={this.childRef.current} text={text} />
```

Comment:
The first time this Tooltip is rendered, `this.childRef.current` is null and even though it's immediately set, it won't trigger another re-render. I get that it's a super rare edge case because we really only need the ref after focus/hover, but it's something that you should consider when designing these sorts of interactions.

Reply:
I agree and I followed current implementation practices with Flyout within Tooltip. The popup mini-component here checks if its null before further rendering. The only way around this would be to trigger a re-render by saving the ref to state for example but since its only appearing on focus/hover I think its safe to follow the practice of Flyout and keep as is like this